### PR TITLE
initial upload of yum_versionlock

### LINKED ISSUE #13086: yum: honour package version pinning via yum-plugin-versionlock
#### Issue Type
Feature Idea

#### Component Name
yum

#### Ansible Version
ansible 1.9.4

#### Ansible Configuration

N/A

#### Environment
systems using yum with _yum-plugin-versionlock_

ansible_pkg_mgr: yum

 yum {{opts}} {{comnmand}}
Loaded **plugins:** langpacks, **versionlock**

#### Summary

This 'feature' would allow one to ensure desired package version is installed AND ensure that the version is not changed outside of the current play (i.e. general system maintenance windows).

The state=latest option is less desirable due to the overhead of needing to update caches for repoquery.  Working with state=installed could ensure that the desired version is installed during play itself.  Exposing a new state would add undue complexity esp where default handling should be present.

yum state=install ignores versionlock settings for matching package.  This appears to be due to the yum module checking if the package is installed though not if it is correct

#### Steps To Reproduce

a) install old version of a package
b) ensure newer package is avaialbe to system via yum/rhn/subscription_manager
c) update /etc/yum/pluginconf.d/versionlock.list with NVR of newer package
d) run play against host using 'yum: name={{package_name}} state=installed

Expected Results:

If the package is already installed, 
and the package matches entry in /etc/yum/pluginconf.d/versionlock.list,
and versionlock plugin is available:

replace states 'installed' and 'latest' with a check
if not matching nvr in file upgrade to desired version(changed)
if it that fails (error)
else (ok)

#### Actual Results
e) observe that the package is not updated to the version locked version.

### PR DESCRIPTION
**Pull request by gyptazy**
Adds support to lock packages in their version to prevent these from updating by yum.

##### SUMMARY
This module adds support to lock packages in their version to prevent these from updating by yum.

Fixes #13086 

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
- yum_versionlock

##### ANSIBLE VERSION
```
ansible 2.4.2.0
```

##### ADDITIONAL INFORMATION
An example run in local playbook

```
TASK [version_lock : lock nginx to prevent updates] **********************************************************************************************************
changed: [localhost]

TASK [version_lock : update nginx] ***************************************************************************************************************************
ok: [localhost]

TASK [version_lock : unlock nginx from updates] **************************************************************************************************************
changed: [localhost]

TASK [version_lock : update nginx] ***************************************************************************************************************************
changed: [localhost]
```

### CODE REVIEW

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: yum_versionlock
+version_added: 2.7
+short_description: Locks/prevents an installed package from beeing updates by (yum) package mamanger.
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: Add `Locks/Unlocks` - as both operations are allowed

Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: s/beeing/being/

Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: `s/mamanger/manager/`

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+description:
+     - This module adds installed packages to yum versionlock to prevent it from beeing updated.
+options:
+  package:
+    description:
+      - A package name like C(httpd).
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: Can we have a list of packages which can be locked/unlocked? just like yum module does.

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  state:
+    description:
+      - Whether to lock (C(present) or unlock (C(absent) a package.
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: (C(present) -> C(present)
(C(absent) -> C(absent)

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    choices: [ present, absent ]
+# informational: requirements for nodes
+requirements:
+- yum
+- yum-versionlock
+author:
+    - Florian Paul Hoberg <florian.hoberg@credativ.de>
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: add github handle too.

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    description: state of used package
+    returned: everytime
+    type: string
+    sample: present
+'''
+
+import os.path
+from ansible.module_utils.basic import AnsibleModule
+
+YUM_BINARY = "/bin/yum"
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: use this instead

```
module.get_bin_path('yum', required=True)
```

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+
+def get_state_yum_versionlock():
+    """ Check for yum plugin dependency """
+    state = os.path.exists("/etc/yum/pluginconf.d/versionlock.conf")
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: Are there any other way to check if version lock plugin is installed ?

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    return state
+
+
+def get_versionlock_packages(module):
+    """ Get an overview of all packages on yum versionlock """
+    rc_code, out, err = module.run_command("%s -q versionlock list"
+                                           % (YUM_BINARY))
+    if rc_code is 0:
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]:
```
rc_code == 0
```

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        return changed
+    else:
+        module.fail_json(msg="Error: " + str(err) + str(out))
+
+
+def main():
+    """ start main program to add/remove a package to yum versionlock"""
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(required=True, type='str', choices=['present', 'absent']),
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: state can be optional with default value as 'present'. Also, default argument type is str so no need to specify explicitly.

File: lib/ansible/modules/packaging/os/yum_versionlock.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: yum_versionlock
+version_added: 2.7
+short_description: Locks/prevents an installed package from beeing updates by (yum) package mamanger.
+description:
+     - This module adds installed packages to yum versionlock to prevent it from beeing updated.
```
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/yum_versionlock.py]: s/beeing/being/

**Review by Akasurde — CHANGES_REQUESTED** [review]
Could you please check inline comments and also add testcase for this module ? thanks.

**Comment by gyptazy** [comment]
ok thanks, unit test file will be created soon.
@Akasurde changes are done.

**Comment by avarins** [comment]
What is the status?
Does a ready_for_review have to be attached?

**Comment by corbinstuard** [comment]
This looks good, can it be reviewed?

**Comment by a-nldisr** [comment]
@Akasurde any plans to check this, would be great to have! :)

**Comment by riemers** [comment]
Want this the Ansible way, would need this too. But since its 2 years old, i don't think i should wait on it. (which is a shame) i could even look into it if that would be the default reply (as far as my knowledge goes)

**Comment by Akasurde** [comment]
@riemers @florianpaulhoberg All community modules are migrated to https://github.com/ansible-collections/community.general repo. Could you please migrate this PR to above-mentioned repo. I am sorry that I missed previous ping messages. Thanks.

**Comment by riemers** [comment]
I moved it over as is, we shall see how far that gets us :)

**Comment by Akasurde** [comment]
@riemers Thanks for migrating this module to new repo.

Migrated to https://github.com/ansible-collections/community.general/pull/324

**Pull request closed without merging**